File: Crio/crio_programming_interview_problems-master/BooleanEvaluation/BooleanEvaluation.cpp
#include<bits/stdc++.h>
using namespace std;

//CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
long long mod=1e9+7;
long long dp[1001][1001][3];
//CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

long long booleanEvaluation(string expression,bool result)
{
    //CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    long long i,j,k,n=expression.length();
    for(k=1;k<=n;k+=2)
    {
        for(i=0;i<n;i+=2)
        {
            if(k==1)
            {
                if(expression[i]=='1')
                {
                    dp[i][i][1]=1;
                    dp[i][i][0]=0;
                }
                else
                {
                    dp[i][i][1]=0;
                    dp[i][i][0]=1;                }
            }
            else
            {
                for(j=i;j<i+k;j+=2)
                {
                    if(expression[j+1]=='&')
                    {
                        dp[i][i+k-1][0]=(dp[i][i+k-1][0]%mod + ((dp[i][j][0]%mod)*((dp[j+2][i+k-1][0]+dp[j+2][i+k-1][1])%mod))%mod + ((dp[i][j][1]%mod)*(dp[j+2][i+k-1][0]%mod))%mod)%mod;
                        dp[i][i+k-1][1]=(dp[i][i+k-1][1]%mod + ((dp[i][j][1]%mod)*(dp[j+2][i+k-1][1]%mod))%mod)%mod;
                    }
                    if(expression[j+1]=='|')
                    {
                        dp[i][i+k-1][0]=(dp[i][i+k-1][0]%mod + ((dp[i][j][0]%mod)*(dp[j+2][i+k-1][0]%mod))%mod)%mod;
                        dp[i][i+k-1][1]=(dp[i][i+k-1][1]%mod + ((dp[i][j][1]%mod)*((dp[j+2][i+k-1][0]+dp[j+2][i+k-1][1])%mod))%mod + ((dp[i][j][0]%mod)*(dp[j+2][i+k-1][1]%mod))%mod)%mod;
                    }
                    if(expression[j+1]=='^')
                    {
                        dp[i][i+k-1][0]=(dp[i][i+k-1][0]%mod + ((dp[i][j][0]%mod)*(dp[j+2][i+k-1][0]%mod))%mod + ((dp[i][j][1]%mod)*(dp[j+2][i+k-1][1]%mod))%mod)%mod ;
                        dp[i][i+k-1][1]=(dp[i][i+k-1][1]%mod + ((dp[i][j][1]%mod)*(dp[j+2][i+k-1][0]%mod))%mod + ((dp[i][j][0]%mod)*(dp[j+2][i+k-1][1]%mod))%mod)%mod;
                    }
                }
            }
        }
    }
    return dp[0][n-1][result]%mod;
    //CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
}
int main()
{
    string s;
    cin>>s;
    bool result;
    cin>>result;
    cout<<booleanEvaluation(s,result);
}

File: Crio/crio_programming_interview_problems-master/BooleanEvaluation/BooleanEvaluation.py
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
MOD = 10 ** 9 + 7
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def booleanEvaluation(expression, result):
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    n = len(expression)
    dp = [[[0,0] for i in range(n)] for j in range(len(expression))]

    numberOfWays = 0
    for left in range(n-1, -1, -2):
        for right in range(left, n, 2):
            for value in range(2):
                if left == right:
                    dp[left][right][value] = int(value == int(expression[left]))
                    continue
                for mid in range(left, right, 2):
                    operator = expression[mid+1]
                    for leftResult in range(2):
                        for rightResult in range(2):
                            if eval('{}{}{}'.format(leftResult, operator, rightResult)) == value:
                                count = (dp[left][mid][leftResult] * dp[mid+2][right][rightResult]) % MOD
                                dp[left][right][value] = (dp[left][right][value] + count) % MOD

    return dp[0][n-1][result]
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

expression = input().strip()
result = int(input())

result = booleanEvaluation(expression, result)
print(result)
